Sources/SwiftFusion/Inference/PenguinExtensions.swift

```diff
@@ -0,0 +1,55 @@
+// TODO: Move all these to Penguin.
+
+import PenguinStructures
+
+extension AnyArrayBuffer {
+  /// Appends `x`, returning the index of the appended element.
+  ///
+  /// - Requires: `self.elementType == T.self`.
+  mutating func unsafelyAppend<T>(_ x: T) -> Int {
+    unsafelyMutate(assumingElementType: Type<T>()) { $0.append(x) }
+  }
+}
+
+/// An `AnyArrayBuffer` dispatcher for elements that are not statically known to support any
+/// operations.
+// We can't use `AnyObject` for this because the bitcast at [1] fails with "Swift runtime failure:
+// Can't unsafeBitCast between types of different sizes" when trying to cast an `AnyArrayBuffer`
+// of some other dispatch to an `AnyArrayBuffer<AnyObject>`.
+//
+// [1] https://github.com/saeta/penguin/blob/47980eb5630ebfd7bc2a88a4a7a60a050dd01b0a/Sources/PenguinStructures/AnyArrayBuffer.swift#L64
```

Comment:
Something's not right here.  The `MemoryLayout<AnyObject.Type>.size` is the same as the size of any class type.  You should put some print statements in there to see what it thinks the types and sizes are.

Some weirdness in this area:
https://bugs.swift.org/browse/SR-13055

Reply:
```
  /// Creates an instance containing the same elements as `src`.
  ///
  /// - Requires: `src.dispatch is Dispatch.Type`.
  public init<OtherDispatch>(unsafelyCasting src: AnyArrayBuffer<OtherDispatch>)
  {
    assert(src.dispatch is Dispatch.Type)
    print("Dispatch.Type: \(Dispatch.Type.self) \(MemoryLayout<Dispatch.Type>.size)")
    print("OtherDispatch.Type: \(OtherDispatch.Type.self) \(MemoryLayout<OtherDispatch.Type>.size)")
    self.storage = src.storage
    self.dispatch = unsafeBitCast(src.dispatch, to: Dispatch.Type.self)
  }
```

Result with `-O`:
```
Dispatch.Type: AnyObject.Protocol 8
OtherDispatch.Type: DifferentiableArrayDispatch.Type 8
... execution proceeds normally ...
```

Result with `-O -cross-module-optimization`:
```
Dispatch.Type: AnyObject.Protocol 0
OtherDispatch.Type: DifferentiableArrayDispatch.Type 8
Illegal instruction
```

Is this a bug in swiftc? I don't know enough about what is supposed to happen to tell.

Reply:
Yes, most definitely a bug when the size of `AnyObject.Protocol` changes based on optimization.

Reply:
filed https://bugs.swift.org/browse/SR-13061